connection/connection_test.go

```diff
@@ -0,0 +1,198 @@
+	// to appear, because then we would have to let the test run
+	// forever.... What we can test is that it returns shortly
+	// after the server appears.
+	startTime := time.Now()
+	var startTimeServer time.Time
+	var stopServer func()
+	var wg sync.WaitGroup
+	wg.Add(1)
+	defer func() {
+		wg.Wait()
+		stopServer()
+	}()
+	// Here we pick a relatively long delay before we start the
+	// server.  If gRPC did go into an exponential backoff before
+	// retrying the connection attempt, then it probably would
+	// not react promptly to the server becoming ready. Currently
+	// it looks like gRPC tries to connect once per second, with
+	// no exponential backoff.
+	delay := 10 * time.Second
+	go func() {
+		defer wg.Done()
+		t.Logf("sleeping %s before starting server", delay)
+		time.Sleep(delay)
+		startTimeServer = time.Now()
+		_, stopServer = startServer(t, tmp)
+	}()
+	conn, err := Connect(path.Join(tmp, serverSock))
+	if assert.NoError(t, err, "connect via absolute path") {
+		endTime := time.Now()
+		assert.NotNil(t, conn, "got a connection")
+		assert.Equal(t, connectivity.Ready, conn.GetState(), "connection ready")
+		if assert.InEpsilon(t, 1*time.Second, endTime.Sub(startTimeServer), 5, "connection established shortly after server starts") {
+			assert.InEpsilon(t, delay, endTime.Sub(startTime), 1)
+		}
+		err = conn.Close()
+		assert.NoError(t, err, "closing connection")
+	}
+}
+
+func TestTimout(t *testing.T) {
+	tmp := tmpDir(t)
+	defer os.RemoveAll(tmp)
+
+	startTime := time.Now()
+	timeout := 5 * time.Second
+	conn, err := Connect(path.Join(tmp, "no-such.sock"), grpc.WithTimeout(timeout), grpc.WithBlock())
+	endTime := time.Now()
+	if assert.Error(t, err, "connection should fail") {
+		assert.InEpsilon(t, timeout, endTime.Sub(startTime), 1, "connection timeout")
+	} else {
+		err := conn.Close()
+		assert.NoError(t, err, "closing connection")
+	}
+}
+
+func TestMethodCall(t *testing.T) {
+	tmp := tmpDir(t)
+	defer os.RemoveAll(tmp)
+	addr, stopServer := startServer(t, tmp)
+	defer stopServer()
+
+	conn, err := Connect(addr)
+	if assert.NoError(t, err, "connect via absolute path") &&
+		assert.NotNil(t, conn, "got a connection") {
+		defer conn.Close()
+		assert.Equal(t, connectivity.Ready, conn.GetState(), "connection ready")
+
+		if err := conn.Invoke(context.Background(), "/connect.v0.Test/Ping", nil, nil); assert.Error(t, err) {
+			errStatus, _ := status.FromError(err)
+			assert.Equal(t, codes.Unimplemented, errStatus.Code(), "not implemented")
+		}
+
+		stopServer()
+		startTime := time.Now()
+		if err := conn.Invoke(context.Background(), "/connect.v0.Test/Ping", nil, nil); assert.Error(t, err) {
+			endTime := time.Now()
+			errStatus, _ := status.FromError(err)
+			assert.Equal(t, codes.Unavailable, errStatus.Code(), "connection lost")
+			assert.InEpsilon(t, time.Second, endTime.Sub(startTime), 1, "connection loss should be detected quickly")
+		}
+
+		// No reconnection either.
```

Comment:
I am not sure what this test tries to do. If server is gone, gRPC should return `codes.Unavailable` quickly. That's correct. But it will try to reconnect if you re-start the server. So `No reconnection either` comment is quite confusing.